### ISSUE TITLE
Questions about the use of softmax in the training process and evaluation process

### ISSUE THREAD

**Issue by ALLLLecy** [cs-mshah/CDUL#2]
During the model training process, the KL divergence between the pseudo-labels and the prediction results was used, which requires us to perform softmax on the prediction results.
```python
            imgs = imgs.to(self.device)
            pseudo_labels = pseudo_labels.to(self.device)
            target_labels = target_labels.to(self.device)
            
            preds = F.softmax(self.model(imgs), dim=-1)
            self.train_metric.update(preds, target_labels)
            self.pseudo_metric.update(pseudo_labels, target_labels)
            kl_loss = self.loss(preds, pseudo_labels)
```
But as far as I know, the commonly used losses in multi-label classification are Binary Cross-Entropy Loss or Asymmetric Loss. Is it reasonable to use softmax here?
In this case, can the loss be changed to the following version
```python
            preds = torch.sigmoid(self.model(imgs))  # [batch_size, num_classes]
            pseudo_labels = pseudo_labels               # [batch_size, num_classes] 
            
            kl_loss = 0
            for i in range(self.num_classes):
                p_pred = torch.stack([preds[:, i], 1 - preds[:, i]], dim=1)                                   # [batch_size, 2]
                p_pseudo = torch.stack([pseudo_labels[:, i], 1 - pseudo_labels[:, i]], dim=1)  # [batch_size, 2]
                
                kl_loss += self.loss(p_pred.log(), p_pseudo) 
            kl_loss /= self.num_classes
```

Furthermore, during the verification process, softmax was also used instead of sigmoid
```python
 def validation(self, epoch: int) -> float:
        """calculate validation mAP"""
        self.model.eval()
        for imgs, targets in tqdm(self.val_dataloader, desc=f"[epoch: {epoch+1}] Validation batch"):
            imgs = imgs.to(self.device)
            targets = targets.to(self.device)
            preds = F.softmax(self.model(imgs), dim=-1)
            self.val_metric.update(preds, targets)

        AP_per_class = self.val_metric.compute()
        mAP = torch.mean(torch.tensor(list(AP_per_class.values())))
        return mAP.item()
```

**Comment by cs-mshah** [comment]
Yes, you are right. I think we have made a slight mistake. Can you train the model with the cache and create a PR with the updated results that you get?

**Comment by ALLLLecy** [comment]
@cs-mshah 
Yes, I see! The key issue lies in the temperature of the CLIP model. In your Settings, it is set to the temperature of the CLIP model itself, which is approximately 4.0. However, this will lead to the probabilities of the models after softmax mostly being around 0.5, making it difficult to distinguish them easily. Refer to OPENCLIP. I set the temperature to 0.01 to widen the gap between the probabilities of each category, and then initialized the cache and trained according to the code above. The performance of CDUL is quite good. After about 20 rounds of training using the strategy in the paper, the train mAP is approximately 0.85 and the val mAP is approximately 0.84.
However, during the training process, I found that the mAP of the pseudo-label dropped rapidly. The subsequent fine-tuning strategy can be carried out according to this
![Image](https://github.com/user-attachments/assets/73d19d0e-c066-4b01-8f3c-b63f8e0374e3)

**Comment by cs-mshah** [comment]
@ALLLLecy Thanks for the insights. Would you be able to raise a PR with your findings and changes? We can merge them so that others can benefit from the improvements.

**Comment by ALLLLecy** [comment]
@cs-mshah 
![Image](https://github.com/user-attachments/assets/1468be1d-0eb6-49ca-b8a2-34d2005bf6f6)
![Image](https://github.com/user-attachments/assets/e233da52-8483-4990-93f0-4420580c9b4c)
![Image](https://github.com/user-attachments/assets/a4589d0c-9d4e-44a4-bdd3-a9ab15545588)
![Image](https://github.com/user-attachments/assets/de4a97f3-83ff-4c4b-af9b-855d44b98b3c)
![Image](https://github.com/user-attachments/assets/d498105c-1151-450d-ad03-59d2216bbee0)
![Image](https://github.com/user-attachments/assets/11f38652-a8e2-4be1-99e1-8b071b4a4ac0)
![Image](https://github.com/user-attachments/assets/0a2cf538-8bf8-4def-952e-7ddd65aad3aa)
![Image](https://github.com/user-attachments/assets/4550f29e-9f66-406b-8413-ef4d98ca234e)
![Image](https://github.com/user-attachments/assets/aa14f5a1-aa7d-484f-b600-e67cd839e916)
![Image](https://github.com/user-attachments/assets/f6873e89-c269-433a-814a-23a688d73112)
![Image](https://github.com/user-attachments/assets/d3f4a60c-024a-4b29-a26c-68d8c85bd8a3)
![Image](https://github.com/user-attachments/assets/b88e3b76-79be-440a-8c68-c86b4cbd154e)
![Image](https://github.com/user-attachments/assets/e307c71f-4827-4cac-9bba-4a98c017be43)
![Image](https://github.com/user-attachments/assets/f353d938-c7fb-4ba4-b133-6d0ac417d224)
![Image](https://github.com/user-attachments/assets/4ee76823-5161-4fc3-8aa4-a81b95507e99)
![Image](https://github.com/user-attachments/assets/4c522ac8-d5de-434a-af86-1742c68cf253)
![Image](https://github.com/user-attachments/assets/d0e3240b-0502-4f63-b8aa-be3bec84c965)
![Image](https://github.com/user-attachments/assets/2ccf2e80-9ff8-4b90-b1a7-3c5590f19927)
![Image](https://github.com/user-attachments/assets/8c854217-0afc-4f55-9073-3fef902fb4eb)
![Image](https://github.com/user-attachments/assets/ea293168-a314-40a5-acee-573d53d5f220)

The PR curve is shown above, and the following is the aggregated data

![Image](https://github.com/user-attachments/assets/66afca8d-e203-4df0-a829-500d74e7f0f4)
